src/main/java/com/mraof/minestuck/entity/AnimatedPathfinderMob.java

```diff
@@ -0,0 +1,107 @@
+package com.mraof.minestuck.entity;
+
+import com.mraof.minestuck.entity.animation.MobAnimation;
+import net.minecraft.network.syncher.EntityDataAccessor;
+import net.minecraft.network.syncher.EntityDataSerializers;
+import net.minecraft.network.syncher.SynchedEntityData;
+import net.minecraft.world.entity.EntityType;
+import net.minecraft.world.entity.PathfinderMob;
+import net.minecraft.world.entity.ai.attributes.AttributeInstance;
+import net.minecraft.world.entity.ai.attributes.AttributeModifier;
+import net.minecraft.world.entity.ai.attributes.Attributes;
+import net.minecraft.world.level.Level;
+
+import java.util.UUID;
+
+/**
+ * Abstract class for handling the actions of Geckolib animated mobs. It keeps track of which animation is in use via a MobAnimation and EntityDataAccessor
+ */
+public abstract class AnimatedPathfinderMob extends PathfinderMob
+{
+	/**
+	 * Retains the action from MobAnimation
+	 */
+	private static final EntityDataAccessor<Integer> CURRENT_ACTION = SynchedEntityData.defineId(AnimatedPathfinderMob.class, EntityDataSerializers.INT);
+	
+	private static final UUID FREEZE_MOB_UUID = UUID.fromString("e4c7543e-335c-4217-8d15-25c157e8ce88");
+	private static final AttributeModifier STATIONARY_MOB_MODIFIER = new AttributeModifier(FREEZE_MOB_UUID, "Stationary Animation Modifier", -1, AttributeModifier.Operation.MULTIPLY_TOTAL);
+	
+	private int remainingAnimationTicks; //starts off as the time value stored in mobAnimation when set. Will be set to -1 for looping animations
+	
+	
+	protected AnimatedPathfinderMob(EntityType<? extends AnimatedPathfinderMob> type, Level level)
+	{
+		super(type, level);
+	}
+	
+	@Override
+	protected void defineSynchedData()
+	{
+		super.defineSynchedData();
+		entityData.define(CURRENT_ACTION, MobAnimation.IDLE_ACTION.ordinal());
+	}
+	
+	@Override
+	public void tick()
+	{
+		super.tick();
+		
+		if(remainingAnimationTicks > 0)
+		{
+			remainingAnimationTicks--;
+			if(remainingAnimationTicks == 0)
+				this.endCurrentAction();
+		}
+	}
+	
+	/**
+	 * Gets called when the duration of time associated with an action has run out.
+	 * Default behavior is to set the action to IDLE, but this function can be overridden to add behavior for specific actions.
+	 **/
+	public void endCurrentAction()
+	{
+		unfreezeMob(); //it is assumed that the idle animation does not require the mob to be frozen, this is a second measure to ensure they do not get stuck
```

Comment:
Doesn't seem neccessary since freezing/unfreezing is the first thing that happens in `setCurrentAnimation()`